edb/edgeql-parser/edgeql-parser-python/src/parser.rs

```diff
@@ -110,60 +105,49 @@ fn downcast_tokens<'a>(
     Ok(buf)
 }
 
-fn get_spec(
-    py: Python,
-    spec_filepath: &str,
-) -> Result<&'static (parser::Spec, PyObject), cpython::PyErr> {
+fn get_spec(py: Python) -> Result<&'static (parser::Spec, PyObject), cpython::PyErr> {
     if let Some(x) = PARSER_SPECS.get() {
         return Ok(x);
+    } else {
+        return Err(cpython::PyErr::new::<AssertionError, _>(
+            py,
+            ("grammar spec not loaded",),
+        ));
     }
-
-    // Try to load from file
-    let spec = load_spec(spec_filepath);
-    let productions = load_productions(py, &spec)?;
-
-    PARSER_SPECS.set((spec, productions)).ok();
-    Ok(PARSER_SPECS.get().unwrap())
 }
 
 /// Loads the grammar specification from file and caches it in memory.
 pub fn preload_spec(py: Python, spec_filepath: &PyString) -> PyResult<PyNone> {
-    get_spec(py, &spec_filepath.to_string(py)?)?;
+    if PARSER_SPECS.get().is_some() {
+        return Ok(PyNone);
+    }
+
+    let spec_filepath = spec_filepath.to_string(py)?;
+    let bytes = std::fs::read(spec_filepath.as_ref()).ok().unwrap();
+
+    let spec: parser::Spec = bitcode::deserialize::<parser::SpecSerializable>(&bytes)
+        .unwrap()
+        .into();
+    let productions = load_productions(py, &spec)?;
+
+    let _ = PARSER_SPECS.set((spec, productions));
     Ok(PyNone)
 }
 
 /// Serialize the grammar specification and write it to a file.
 ///
 /// Called from setup.py.
-pub fn save_spec(py: Python, py_spec: &PyObject, dst: &PyString) -> PyResult<PyNone> {
-    let spec = spec_from_python(py, py_spec)?;
-    let spec_serialized = bitcode::serialize(&spec).unwrap();
+pub fn save_spec(py: Python, spec_json: &PyString, dst: &PyString) -> PyResult<PyNone> {
+    let spec_json = spec_json.to_string(py).unwrap();
+    let spec: parser::SpecSerializable = serde_json::from_str(&spec_json).unwrap();
+    let spec_bitcode = bitcode::serialize(&spec).unwrap();
 
     let dst = dst.to_string(py)?.to_string();
 
-    std::fs::write(dst, spec_serialized).ok().unwrap();
+    std::fs::write(dst, spec_bitcode).ok().unwrap();
     Ok(PyNone)
 }
 
-fn load_spec(spec_filepath: &str) -> parser::Spec {
-    let bytes = std::fs::read(spec_filepath).ok().unwrap();
-
-    bitcode::deserialize::<parser::SpecSerializable>(&bytes)
-        .unwrap()
-        .into()
-}
-
-fn spec_from_python(py: Python, py_spec: &PyObject) -> PyResult<parser::SpecSerializable> {
-    let spec_to_json = py.import("edb.common.parsing")?.get(py, "spec_to_json")?;
-
-    let res = spec_to_json.call(py, (py_spec,), None)?;
-
-    let spec_json = PyString::downcast_from(py, res)?;
-    let spec_json = spec_json.to_string(py).unwrap();
-
-    Ok(serde_json::from_str::<parser::SpecSerializable>(&spec_json).unwrap())
-}
-
 fn load_productions(py: Python<'_>, spec: &parser::Spec) -> PyResult<PyObject> {
```

Comment:
This isn't a blocker, but maybe we could make `preload_spec` return the list of production/classname tuples, and call `load_spec_productions` from python when we call `preload_spec` and cache that side of it in python?

Then there wouldn't be any more tight coupling between the python side and the rust side.

Reply:
That a bit of work for marginal benefit - rust side is already quite decoupled, since only `edgeql-parser-python` crate depends on Python.